src/H5Z-SPERR/src/h5z-sperr.c

```diff
@@ -0,0 +1,347 @@
+    printf("%s: %d, real_dims = %d\n", __FILE__, __LINE__, real_dims);
+#endif
+    H5Epush(H5E_DEFAULT, __FILE__, __func__, __LINE__, H5E_ERR_CLS, H5E_PLINE, H5E_BADTYPE,
+            "bad chunk dimensions: only true 2D slices or 3D volumes are supported in H5Z-SPERR");
+    return 0;
+  }
+
+  /* Real chunk dimensions must be at least 9 */
+  for (int i = 0; i < ndims; i++) {
+    if (chunks[i] > 1 && chunks[i] < 9) {
+#ifndef NDEBUG
+      printf("%s: %d, chunks[%d] = %llu\n", __FILE__, __LINE__, i, chunks[i]);
+#endif
+      H5Epush(H5E_DEFAULT, __FILE__, __func__, __LINE__, H5E_ERR_CLS, H5E_PLINE, H5E_BADTYPE,
+              "bad chunk dimensions: any dimension must be at least 9. (may relax this requirement "
+              "in the future)");
+      return 0;
+    }
+  }
+
+  return 1;
+}
+
+/*
+ * Pack information about the input data into an `unsigned int`.
+ * It returns the encoded unsigned int, which shouldn't be zero.
+ */
+static unsigned int H5Z_SPERR_pack_data_type(int rank,  /* Input */
+                                             int dtype) /* Input */
+{
+  unsigned int ret = 0;
+
+  /*
+   * Bit position 0-3 to encode the rank.
+   * Since this function is called from `set_local()`, it should always be 2 or 3.
+   */
+  if (rank == 2) {
+    ret |= 1u << 1; /* Position 1 */
+  }
+  else {
+    assert(rank == 3);
+    ret |= 1u;      /* Position 0 */
+    ret |= 1u << 1; /* Position 1 */
+  }
+
+  /*
+   * Bit position 4-7 encode data type.
+   * Only float (1) and double (0) are supported right now.
+   */
+  if (dtype == 1)   /* is_float   */
+    ret |= 1u << 4; /* Position 4 */
+  else
+    assert(dtype == 0);
+
+  return ret;
+}
+
+/*
+ * Unpack information about the input data from an `unsigned int`.
+ */
+static void H5Z_SPERR_unpack_data_type(unsigned int meta, /* Input  */
+                                       int* rank,         /* Output */
+                                       int* dtype)        /* Output */
+{
+  /*
+   * Extract rank from bit positions 0-3.
+   */
+  unsigned pos0 = meta & 1u;
+  unsigned pos1 = meta & (1u << 1);
+  if (!pos0 && pos1)
+    *rank = 2;
+  else if (pos0 && pos1)
+    *rank = 3;
+  else { /* error */
+    H5Epush(H5E_DEFAULT, __FILE__, __func__, __LINE__, H5E_ERR_CLS, H5E_PLINE, H5E_BADVALUE,
+            "Rank is not 2 or 3.");
+  }
+
+  /*
+   * Extract data type from position 4-7.
+   * Only float and double are supported right now.
+   */
+  unsigned pos4 = meta & (1u << 4);
+  if (pos4)
+    *dtype = 1; /* is_float  */
+  else
+    *dtype = 0; /* is_double */
+}
+
+static herr_t H5Z_set_local_sperr(hid_t dcpl_id, hid_t type_id, hid_t space_id)
+{
+  /*
+   * 	dcpl_id	  Dataset creation property list identifier
+   * 	type_id	  Datatype identifier
+   * 	space_id	Dataspace identifier
+   */
+
+  /* Get the user-specified compression mode and quality. */
+  size_t user_cd_nelem = 2, nchar = 16;
+  unsigned int user_cd_values[user_cd_nelem], flags;
+  char name[nchar];
```

Comment:
Visual Studio is not happy with this code:
```
src/H5Z-SPERR/src/h5z-sperr.c(161): error C2057: expected constant expression
src/H5Z-SPERR/src/h5z-sperr.c(161): error C2466: cannot allocate an array of constant size 0
src/H5Z-SPERR/src/h5z-sperr.c(161): error C2133: 'user_cd_values': unknown size
src/H5Z-SPERR/src/h5z-sperr.c(162): error C2057: expected constant expression
src/H5Z-SPERR/src/h5z-sperr.c(162): error C2466: cannot allocate an array of constant size 0
src/H5Z-SPERR/src/h5z-sperr.c(162): error C2133: 'name': unknown size
src/H5Z-SPERR/src/h5z-sperr.c(204): warning C4244: '=': conversion from 'hsize_t' to 'unsigned int', possible loss of data
```

This should fix it:
```suggestion
  size_t user_cd_nelem = 2, nchar = 16;
  unsigned int user_cd_values[2], flags;
  char name[16];
```
... or maybe use of `const` :shrug: 

[full log](https://ci.appveyor.com/project/ESRF/hdf5plugin/builds/49866162/job/ro8kdw4am47l6uf5)

Reply:
I have re-worked the code above, so if you do a git pull it should have my latest changes. If you think I should make a new minor release (`v0.1.3` for example) to include this change, I could do it too! 

I'll need to figure out how to run the python tests though :)

Reply:
Thanks for the fix! I pulled the changes

A new minor release would be great, but maybe wait a bit too see if other such small changes are needed while testing hdf5plugin on different platforms.

Here is some documentation on how to run the tests once hdf5plugin is installed (`pip install .` or the old `python setup.py install`): https://hdf5plugin.readthedocs.io/en/stable/contribute.html#testing

To spare compilation time, you can disable the other filters by setting the `HDF5PLUGIN_STRIP` environment variable to `"bzip2,lz4,bshuf,blosc,blosc2,fcidecomp,zfp,sz,sz3,zstd"` when building hdf5plugin.

Reply:
I followed the installation and test instructions, and it shows all tests are successful. I'm not sure what is exactly tested though...

Reply:
This runs basic write/read tests for all available filters with a few different options and easy to compress data.
It is not meant to test the compression libraries itself but rather test that there is no obvious issue: it checks that some compression occurs and that the read-back data is the same (or close to) the original one.

https://github.com/silx-kit/hdf5plugin/blob/af093e455952df5f8aae5d6ced51cb1ee55805a2/src/hdf5plugin/test.py#L264-L275